### ISSUE TITLE
Hard-coding of architecture directory in modulefile paths

### ISSUE THREAD

**Issue by roguephysicist** [LLNL/spack#16486]
The architecture path is hard-coded into the generated modulefile paths:

https://github.com/spack/spack/blob/2cb4ae747d6dba8817f67ae98d78c39883cb2cff/lib/spack/spack/modules/common.py#L570

This is very inconvenient if the user wants a unified modulefile list even if some packages are built with different targets. At least for the Tcl modulefiles, rather than being hard-coded, couldn't this be part of `naming_scheme` (`{architecture}`), and thus easily removed or changed?

**Comment by alalazo** [comment]
> This is very inconvenient if the user wants a unified modulefile list even if some packages are built with different targets

Hmm, the current naming scheme results in something like:
```
/mydir/spack/share/spack/modules/linux-ubuntu18.04-broadwell/python-3.7.6-gcc-9.0.1-sb5ocq4
```
Usually people would:
```console
$ export MODULEPATH=/mydir/spack/share/spack/modules/linux-ubuntu18.04-broadwell
```
but to do what you ask for I think it suffices to:
```console
$ export MODULEPATH=/mydir/spack/share/spack/modules
```
In that way you should see all the module files that have been generated for any target. Would that solve your use case?

**Comment by roguephysicist** [comment]
> Would that solve your use case?

No, because now your modulefile layout looks like this:

```
----------------------------------------------- /mydir/spack/share/spack/modules -------------------------------------------------
linux-rhel7-x86_64/autoconf/2.69-gcc    linux-rhel7-x86_64/libxrender/0.9.10-gcc        linux-rhel7-x86_64/py-ipython-genutils/0.2.0-gcc
linux-rhel7-broadwell/cfitsio/3.450-gcc linux-rhel7-broadwell/py-jupyter-core/4.6.3-gcc linux-rhel7-broadwell/cmake/3.17.1-gcc
```

which is far from optimal. And what if you have modulefiles for system utilities that need to appear alongside the ones generated by Spack? The 'arch' directory prevents from having a single, unified layout.

Furthermore, `{architecture}` can be specified in the `naming_scheme` property, which can be the default behavior for people who want that. With the current scheme, if my `modules.yaml` file is

```yaml
modules:
  tcl:
    naming_scheme: '{architecture}/{name}/{version}-{compiler.name}'
```
then my modulefiles end up in `/mydir/spack/share/spack/modules/linux-rhel7-x86_64/linux-rhel7-x86_64/`, which is not logical.

**Comment by alalazo** [comment]
> which is far from optimal. 

What is optimal?

**Comment by roguephysicist** [comment]
Something that is easy for users, mostly non-computer science researchers, to work with.

**Comment by alalazo** [comment]
> Something that is easy for users, mostly non-computer science researchers, to work with.

Hmm, can you be more specific? Trying to understand what is your use case and which are the limitations you see.

**Comment by roguephysicist** [comment]
Sure thing.

I am using Spack to automate and manage the software installation on our centralized university cluster. On the hardware side, our cluster is comprised of several different microarchitectures (Haswell, Broadwell, Skylake). On the software side, we serve 14 different departments, so there is quite a large variety of packages that need to be installed and maintained (many of these have a lot of dependency overlap).

End-users typically load their software via modulefiles, which can be done inside their Slurm batch scripts. Historically, our modulefiles have always "autoloaded" other modulefiles in order to satisfy their dependencies, and I was happy to see that Spack handles this very well.

My primary toolchain is comprised of the Intel compilers, the Intel MKL, and we are heavily inclined towards using Open MPI due to its support of the faster usNIC interface our nodes have. The Intel tools are loaded via modulefiles which I made separately. I have already integrated two versions of the Intel tools into Spack with great results, and have also used Spack to build Open MPI with support for Slurm and usNIC. I have also built a good standard toolchain for GCC 8.4.0, so Spack has my toolchain needs covered.

I would like to integrate all of our modulefiles into a single, unified layout. This would include modulefiles that are created separately and those that are generated with Spack for several `architectures` (x86_64, Skylake, etc.). I also have [this pull request](https://github.com/spack/spack/pull/16485) (that you are also reviewing) that would allow for simple modulefile categories. 

In my opinion this would provide the best experience for end-users, because we can design an "intuitive" organization (`applications`, `compilers`, `libraries`, etc.) rather than based on relatively arbitrary technical points (`linux`, `ubuntu`, `rhel7`, `haswell`, etc.). This is especially true for non-technical end-users that are most likely unfamiliar with HPC systems and even Linux in general.

**Comment by alalazo** [comment]
@roguephysicist Thanks for the detailed explanation.

> I am using Spack to automate and manage the software installation on our centralized university cluster. [ ... ] so Spack has my toolchain needs covered.

Up to this point I would say this sounds familiar, as I have done the same thing for EPFL since not so long ago. The only thing I would recommend against is:

> The Intel tools are loaded via modulefiles which I made separately.

I understand why one wants to install Intel manually with the current state of Spack, but my advice here would be to register Intel as an external tool in `packages.yaml` and customize `modules.yaml` so that Spack generates the module file you want for this manually installed Intel. 

> I would like to integrate all of our modulefiles into a single, unified layout. This would include modulefiles that are created separately and those that are generated with Spack for several architectures (x86_64, Skylake, etc.).

That's the part I don't fully grasp of your request. How do you want to distinguish between the same software built for different architectures? How should your users load e.g. `gromacs` on a `skylake` node as opposed to a `zen2` node? Asking because the two requests you outline above:

1. Include modulefiles that are created separately
2. Include those that are generated with Spack for several architectures

can be solved by setting `MODULEPATH` appropriately on your machines. Since that doesn't seem a solution that fits your case I have the feeling I'm missing some part of the problem.

> In my opinion this would provide the best experience [ ... ] is especially true for non-technical end-users that are most likely unfamiliar with HPC systems and even Linux in general.

I still need to have a look at that PR. That said it's not mandatory to expose all the complexity to users. For instance one solution we were adopting at EPFL was to design module file trees so that they would look the same on each different type of node you land on:
```
/mydir/spack/share/spack/modules/linux-ubuntu18.04-broadwell/<same-tree-from-here-on>
/mydir/spack/share/spack/modules/linux-centos7-zen2/<same-tree-from-here-on>
```
This ensures that `module load` commands are the same on every node type and puts on sysadmins the burden to setup MODULEPATH correctly depending on the node you land on. The user won't ever have to know details on the node and everything works "transparently" to him.

**Comment by roguephysicist** [comment]
> I understand why one wants to install Intel manually with the current state of Spack, but my advice here would be to register Intel as an external tool in packages.yaml and customize modules.yaml so that Spack generates the module file you want for this manually installed Intel.

This is exactly what I did. The compilers correctly specified in `compilers.yaml` (even with modules), and the remaining libraries and tools are all specified in `packages.yaml`. I have used the Intel MKL from within Spack on many packages with good results.

> That's the part I don't fully grasp of your request. How do you want to distinguish between the same software built for different architectures?

I am not particularly planning on doing several builds of the **same** software for different architectures. Some packages that do not benefit from optimization could just be built with `target=x86_64`. It was my understanding that the recent gen Intel platforms mostly share the same instruction set; my thinking is that something built with `target=haswell` would work fine on a Skylake machine, or is my thought process faulty? We only have Intel hardware.

> Since that doesn't seem a solution that fits your case I have the feeling I'm missing some part of the problem.

Let's assume I have three module directories: one for `linux-rhel7-x86_64`, one for `linux-rhel7-haswell`, and one for my non-Spack modules, `current`. Sure, I could include these in `$MODULEPATH`, but when the user does a `module av`, there are now 3 separate listings that show up. This has often been a source of confusion with our users, especially those who are just learning about what software is available on the cluster.

Lastly, I'll return to the point I made in my first response:

> Furthermore, `{architecture}` can be specified in the `naming_scheme` property, which can be the default behavior for people who want that. With the current scheme, if my `modules.yaml` file is

```yaml
modules:
  tcl:
    naming_scheme: '{architecture}/{name}/{version}-{compiler.name}'
```
> then my modulefiles end up in `/mydir/spack/share/spack/modules/linux-rhel7-x86_64/linux-rhel7-x86_64/`, which is not logical.

My point with this was: Spack allows for a tremendous amount of control. Why hard-code that directory in there, if it can be easily added with `naming_scheme`? Having `{architecture}` in there could the default behavior, but the extra benefit is that someone like me could easily change it if they want a different directory structure.
